src/transport_ofi.c

```diff
@@ -536,63 +632,192 @@ static int populate_mr_tables(void)
         RAISE_ERROR(-1);
     }
 
-    /* Binary OPS check */
-    for(i=0; i<3; i++) {//DT
-      for(j=0; j<3; j++) { //OPS
-        ret = fi_atomicvalid(shmem_transport_ofi_epfd, SHM_DT_INT[i], SHM_BOPS[j],
-                        &atomic_size);
-        if(ret!=0 || atomic_size == 0) {
-           OFI_ERRMSG("ret=%d atomic_size=%d %d %d\n", ret, (int)atomic_size, i, j);
-	   return ret;
-        }
-      }
-    }
+    /* Standard OPS check */
+    ret = atomicvalid_DTxOP(SIZEOF_AMO_DT, SIZEOF_AMO_OPS, DT_AMO_STANDARD,
+                      AMO_STANDARD_OPS, atomicwarn);
+    if(ret)
+        return ret;
 
-    /* OTHER OPS check */
-    for(i=0; i<5; i++) {//DT
-      for(j=0; j<4; j++) { //OPS
-        ret = fi_atomicvalid(shmem_transport_ofi_epfd, SHM_DT_CMP[i], SHM_OPS[j],
-                        &atomic_size);
-        if(ret!=0 || atomic_size == 0) {
-           OFI_ERRMSG("ret=%d atomic_size=%d %d %d\n", ret, (int)atomic_size, i, j);
-	   return ret;
-        }
+    ret = fetch_atomicvalid_DTxOP(SIZEOF_AMO_DT, SIZEOF_AMO_FOPS,
+                    DT_AMO_STANDARD, FETCH_AMO_STANDARD_OPS, atomicwarn);
+    if(ret)
+        return ret;
+
+    ret = compare_atomicvalid_DTxOP(SIZEOF_AMO_DT, SIZEOF_AMO_COPS,
+                    DT_AMO_STANDARD, COMPARE_AMO_STANDARD_OPS, atomicwarn);
+    if(ret)
+        return ret;
+
+    /* Extended OPS check */
+    ret = atomicvalid_DTxOP(SIZEOF_AMO_EX_DT, SIZEOF_AMO_EX_OPS, DT_AMO_EXTENDED,
+                      AMO_EXTENDED_OPS, atomicwarn);
+    if(ret)
+        return ret;
+
+    ret = fetch_atomicvalid_DTxOP(SIZEOF_AMO_EX_DT, SIZEOF_AMO_EX_FOPS,
+                    DT_AMO_EXTENDED, FETCH_AMO_EXTENDED_OPS, atomicwarn);
+    if(ret)
+        return ret;
+
+    /* Reduction OPS check */
+    ret = atomicvalid_DTxOP(SIZEOF_RED_DT, SIZEOF_RED_OPS, DT_REDUCE_BITWISE,
+                      REDUCE_BITWISE_OPS, atomicwarn);
+    if(ret)
+        return ret;
+
+    ret = atomicvalid_DTxOP(SIZEOF_REDC_DT, SIZEOF_REDC_OPS, DT_REDUCE_COMPARE,
+                      REDUCE_COMPARE_OPS, atomicwarn);
+    if(ret)
+        return ret;
+
+    ret = atomicvalid_DTxOP(SIZEOF_REDA_DT, SIZEOF_REDA_OPS, DT_REDUCE_ARITH,
+                      REDUCE_ARITH_OPS, atomicwarn);
+    if(ret)
+        return ret;
+
+    /* Internal atomic requirement */
+    ret = compare_atomicvalid_DTxOP(SIZEOF_INTERNAL_REQ_DT, SIZEOF_INTERNAL_REQ_OPS,
+                    DT_INTERNAL_REQ, INTERNAL_REQ_OPS, atomicwarn);
+    if(ret)
+        return ret;
+
+    /* LONG DOUBLE limitation is common */
+    for(j=0; j<SIZEOF_REDC_OPS; j++) { //OPS
+      ret = fi_atomicvalid(shmem_transport_ofi_epfd, SHM_INTERNAL_LONG_DOUBLE,
+                REDUCE_COMPARE_OPS[j], &atomic_size);
+      if(ret!=0 || atomic_size == 0) {
+	    shmem_transport_have_long_double = 0;
+		break;
+	  } else if((atomic_size*sizeof(long double)) !=
+                                    shmem_transport_ofi_max_atomic_size) {
```

Comment:
Is this check new?  I'm not sure I understand why the long double support would have to be equal to (e.g. not less than) ```max_atomic_size```?